### ISSUE TITLE
bug: Neptune GremlinServer 3.7.1 Cardinality issue

### ISSUE THREAD

**Issue by tsanton** [localstack/localstack#11323]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Running the following MergeV-mutation:

```
g.mergeV([(T.id): "1"]).
    option(Merge.onCreate, [(T.label): 'Person', name: single('marko'), age: single(29)]).
    option(Merge.onMatch, [age: single(30)]);
```

Generates this insert (based `g.V().properties()`) :

```
[
  {
    "key": "name",
    "value": {
      "sourceInstructions": [
        {
          "operator": "CardinalityValueTraversal",
          "arguments": [
            "single",
            "marko"
          ]
        }
      ],
      "stepInstructions": [],
      "bindings": {},
      "instructions": [
        {
          "operator": "CardinalityValueTraversal",
          "arguments": [
            "single",
            "marko"
          ]
        }
      ],
      "empty": false
    },
      },
  {
    "key": "age",
    "value": {
      "sourceInstructions": [
        {
          "operator": "CardinalityValueTraversal",
          "arguments": [
            "single",
            29
          ]
        }
      ],
      "stepInstructions": [],
      "bindings": {},
      "instructions": [
        {
          "operator": "CardinalityValueTraversal",
          "arguments": [
            "single",
            29
          ]
        }
      ],
      "empty": false
    },
      }
]
```

### Expected Behavior

I expect the `MergeV` and g.V().properties() to produce the following output:

```
[
  { "key": "name", "value": "marko" },
  { "key": "age",    "value": 29 }
]
```

I believe the issue is due to [this](https://github.com/apache/tinkerpop/blob/master/CHANGELOG.asciidoc#tinkerpop-372-april-8-2024) fix that came in the Tinkerpop 3.7.2 release: `Fixed bug in onCreate for mergeV() where use of the Cardinality functions was not properly handled`

My question is, how does this query handle in AWS Neptune? Have AWS implemented the same bug, or have they fixed it on their side?

### How are you starting LocalStack?

Custom (please describe below)

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

DotNet Testcontainers, but it's equivalent to `docker run localstack/localstack-pro:latest` with a few environment variables.

### Environment

Inspecting the docker image gives me the following env-variables.
"LOCALSTACK_BUILD_DATE=2024-07-01",
  "LOCALSTACK_BUILD_GIT_HASH=c9a014a",
  "LOCALSTACK_BUILD_VERSION=3.5.1.dev",

### Anything else?

When you first released your preview of Neptune transactions one could override the the default Gremlin Server (3.6.2 I believe it was) and Engine Versions setup by piping in the environment variable `NEPTUNE_ENABLE_TRANSACTION` to have you override and use the Tinkerpop 3.7.2 binaries behind the scenes. I believe this was for Localstack Pro 3.5.1 as it was my last wrapped build. Not sure if something equivalent can be enabled to let us override with 3.7.2, somewhat depending on how AWS Neptune handles the above mentioned merge. 

I don't have access to my own instance of Neptune just yet, so hoping you guys can check it out and get back to me

/T

**Comment by tsanton** [comment]
Hi again, @cloutierMat!

Though I'd shamelessly tag you in this one hoping you can shed some light on this issue based on your last splendid implementation of transactions :)

Hope all is well!

/T

**Comment by cloutierMat** [comment]
Hi @tsanton,

Thank you for reporting this issue. The expected behavior you are mentioning is, indeed, what we would expect to receive. However I am not able to reproduce the issue using localstack.

Could you please make a working test case and publish it here.

Note:
The bug fix mentioned in the 3.7.2 release enables the syntax `option(Merge.onCreate, [age: single(30)])` as from `<=3.7.1` to pass in cardinality with `mergeV` the following syntax was required `option(Merge.onCreate, sideEffect(property(single,'age', 30)).constant([:]))`. This only affects `onCreate`. see https://issues.apache.org/jira/browse/TINKERPOP-3049

**Comment by tsanton** [comment]
Sorry for the late reply.

So I run this docker compose:

```yaml
services:
  localstack:
    environment:
      AWS_DEFAULT_REGION: eu-west-1
      ACTIVATE_PRO: "1"
      DEBUG: "0"
      DOCKER_HOST: unix:///var/run/docker.sock
      LOCALSTACK_AUTH_TOKEN: ${LOCALSTACK_AUTH_TOKEN}
      LS_LOG: WARN
      SERVICES: iam,events,ecr,lambda,logs,s3,neptune
    image: localstack/localstack-pro:3.6.0
    ports:
      - "4566:4566"
      - "4510:4510"
```

I then `docker exec -it...` into the container and run the following commands:

```bash
awslocal neptune create-db-cluster \
    --engine neptune \
    --engine-version 1.3.2.1 \
    --db-cluster-identifier neptune-cluster

awslocal neptune create-db-instance \
    --db-cluster-identifier neptune-cluster \
    --db-instance-identifier neptune-instance \
    --engine neptune \
    --db-instance-class db.t3.medium
```

Then I use the default credentials & port to access the graph through GdotV, and fire off the the following mutation:

```gremin
g.mergeV([(T.id): "5"]).
    option(Merge.onCreate, [(T.label): 'UnitV', name: single('marko'), age: single(29)]).
    option(Merge.onMatch, [age: single(30)]);
````

Then I query out the properties of all vertexes:

```gremin
g.V().properties()
```

And I get the merge instruction object, not the expected Key:Value:

```json
[
  {
    "key": "name",
    "value": {
      "sourceInstructions": [
        {
          "operator": "CardinalityValueTraversal",
          "arguments": [
            "single",
            "marko"
          ]
        }
      ],
      "stepInstructions": [],
      "bindings": {},
      ......
```

If you're not seing this @cloutierMat, it makes me wonder if it can be the OS? 
I'm on a Mac M2 with Sonoma 14.5.
Naturally then the docker image is linux/arm64. Here is the sha from `docker inspect`: `"Image": "sha256:0e7757c922157e7ce9e269797fb808c169f33ef68f0287550ed9ee94a62e6c85"`

I can also add that `Gremlin.version();` is returning null which I think is a bit odd?

**Comment by tsanton** [comment]
To add some context, in order to "fix" this issue and make my localstack run I'm using the 3.5.0 tag and throwing in the `NEPTUNE_ENABLE_TRANSACTION: "1"` environment variable. This runs `mergeV` and `mergeE`  without issues :)

**Comment by cloutierMat** [comment]
That is curious, I can't seem to find what causes this issue with `3.7.1` and why I can't reproduce. I am not familiar with GdotV and how it might transform the query.

I will have a second look again this week, and see that a fix is included in the next release.

**Comment by cloutierMat** [comment]
Hello @tsanton,

I apologise for the delay in fixing this issue. I have been trying to reproduce the issue in order to fix it properly.

A fix has been merged and will be available in our next `latest` nightly build and will be released in the upcoming `3.8` release.

Thank you for your patience and flexibility around using `3.5.0` while I was working on this issue. I will close the issue for now, but feel free to reopen if the problem persists beyond the new release. :pray:
